design/view-change.md

```diff
@@ -0,0 +1,346 @@
+  a simple callback which is called on every message (forcing each subscriber
+  to implement their own routing logic) or as a different callbacks per each
+  message type.
+- `unsubscribe(id)` - unsubscribe from network messages
+- `add_filter(predicate)` - stash or discard all incoming messages based on
+  passed predicate (closure), return unique identifier
+- `remove_filter(id)` - resume processing of all messages that were filtered
+  by previously added predicate. Question is still open whether messages
+  should be discarded or just stashed and re-executed on resume and how
+  interface for this should be implemented.
+
+### Executor
+
+- `validate(batch)` - rewind (if needed) uncommited state to point where batch
+  could be applied, check that batch is valid and apply it. If batch was
+  already commited just accept it as valid. This probably should be split
+  into several methods (like `is_commited`, `revert`, and so on), but this
+  is to be determined during real work on code.
+
+### Orderer
+
+- `view_no()` - return current viewNo
+- `enter_next_view()` - increment current viewNo, discard all previous messages
+- `preprepared()` - return list of pre-prepared batches
+- `prepared()` - return list of prepared batches
+- `preprepare(batch)` - unconditionally put some batch into pre-prepared state
+  producing all necessary side effects (like sending PREPARE messages from
+  backup replicas)
+
+Also it's possible that it will be easier to implement `batches()` method
+that return list of batches in any state and then query them for their state
+(pre-prepared, prepared, commited).
+
+### Checkpointer
+
+- `checkpoints()` - return list of all available checkpoints
+- `stable_checkpoint()` - return current stable checkpoint
+- `set_stable_checkpoint(h)` - update current stable checkpoint
+
+### Viewchanger
+
+- `__init__(network, executor, orderer, checkpointer)` - initialize viewchanger
+  explicitly passing all dependencies
+- `start_view_change()` - initiate view change protocol, called by node (or
+  probably monitor) when we need to perform view change
+- `is_view_change_in_progress()` - check if view change is in progress.
+  Probably there also should be callback on view-change start/stop
+- `process_time(timestamp)` - update internal time, possibly performing some
+  actions. Called periodically by node, probably could be improved in future
+  to use some abstract timer which can be subscribed to and set nearest needed
+  timeout so it won't call tick unnecessarily. Main reason to inject timestamps
+  externally is improved testability.
+- `process_view_change()` - process _VIEW-CHANGE_ message
+- `process_view_change_ack()` - process _VIEW-CHANGE-ACK_ message
+- `process_new_view()` - process _NEW-VIEW_ message
+- `process_status_pending()` - process _STATUS-PENDING_ message
+
+## Implementation plan (minimal)
+
+- enable full ordering of batches that were already ordered, make their
```

Comment:
@ashcherbakov I'm afraid this feature will be used in a context where it's impossible to understand whether batch is new or from previous view unless some field is added to messages, so I believe we should do it always. When we stop resetting ppSeqNo in new view this won't be needed, so it's a temporary solution. The only reason I propose to do it now is that I feel it will be more safe to do on current codebase than the other variant.